File: SASGMACustomEventInterstitial.m
//
//  SASGMACustomEventInterstitial.m
//
//  Created by Julien Gomez on 22/06/16.
//  Copyright © 2018 Smart AdServer. All rights reserved.
//

#import "SASGMACustomEventInterstitial.h"
#import <SASDisplayKit/SASDisplayKit.h>
#import "SASGMACustomEventConstants.h"
#import "SASGMAUtils.h"

NS_ASSUME_NONNULL_BEGIN

@interface SASGMACustomEventInterstitial () <SASInterstitialManagerDelegate>

@property(nonatomic, strong) SASInterstitialManager *interstitialManager;

@end

@implementation SASGMACustomEventInterstitial

@synthesize delegate;

#pragma mark GMACustomEventInterstitial implementation

- (void)requestInterstitialAdWithParameter:(nullable NSString *)serverParameter
                                     label:(nullable NSString *)serverLabel
                                   request:(GADCustomEventRequest *)request {
        
    // Placement parsing from the server string
    SASAdPlacement *adPlacement = [SASGMAUtils placementWithDFPServerParameter:serverParameter request:request];
    
    if (adPlacement != nil) {
        
        // Instantiating an interstitial manager and loading an interstitial from it
        self.interstitialManager = [[SASInterstitialManager alloc] initWithPlacement:adPlacement delegate:self];
        [self.interstitialManager load];
        
    } else {
        
        // Placement is invalid, sending an error
        NSError *error = [NSError errorWithDomain:kSASGMAErrorDomain code:kSASGMAErrorCodeInvalidServerParameters userInfo:nil];
        [self.delegate customEventInterstitial:self didFailAd:error];
        
    }

}

- (void)presentFromRootViewController:(UIViewController *)rootViewController {
    if (self.interstitialManager.adStatus == SASAdStatusReady) {
        // Showing the interstitial if ready
        [self.interstitialManager showFromViewController:rootViewController];
    }
}

#pragma mark SASAdViewDelegate implementation

- (void)interstitialManager:(SASInterstitialManager *)manager didLoadAd:(SASAd *)ad {
    [self.delegate customEventInterstitialDidReceiveAd:self];
}

- (void)interstitialManager:(SASInterstitialManager *)manager didFailToLoadWithError:(NSError *)error {
    [self.delegate customEventInterstitial:self didFailAd:error];
}

- (BOOL)interstitialManager:(SASInterstitialManager *)manager shouldHandleURL:(NSURL *)URL {
    [self.delegate customEventInterstitialWasClicked:self];
    return YES;
}

- (void)interstitialManager:(SASInterstitialManager *)manager didAppearFromViewController:(UIViewController *)viewController {
    [self.delegate customEventInterstitialWillPresent:self];
}

- (void)interstitialManager:(SASInterstitialManager *)manager didDisappearFromViewController:(UIViewController *)viewController {
    [self.delegate customEventInterstitialDidDismiss:self];
}

@end

NS_ASSUME_NONNULL_END

File: SASGMACustomEventInterstitial.swift
//
//  SASGMACustomEventInterstitial.swift
//  NewSample
//
//  Created by Guillaume Laubier on 09/09/2021.
//

import Foundation
import GoogleMobileAds
import SASDisplayKit

@objc(SASGMACustomEventInterstitial)
class SASGMACustomEventInterstitial : NSObject, GADCustomEventInterstitial, SASInterstitialManagerDelegate {
    
    private var interstitialManager: SASInterstitialManager?
    
    var delegate: GADCustomEventInterstitialDelegate?
    
    required override init() {
    }
    
    func requestAd(withParameter serverParameter: String?, label serverLabel: String?, request: GADCustomEventRequest) {
        guard let placement = SASGMAUtils.placementWith(serverParameter: serverParameter, request: request, extras: nil) else {
            // Placement is invalid, sending an error
            let error = NSError(domain: SASGMAUtils.kSASGMAErrorDomain, code: SASGMAUtils.kSASGMAErrorCodeInvalidServerParameters, userInfo: nil)
            self.delegate?.customEventInterstitial(self, didFailAd: error)
            return
        }
        
        // Instantiating an SASInterstitialManager
        interstitialManager = SASInterstitialManager(placement: placement, delegate: self)
        
        // Load the interstitial
        interstitialManager?.load()
    }
    
    func present(fromRootViewController rootViewController: UIViewController) {
        if (interstitialManager?.adStatus == SASAdStatus.ready) {
            // Showing the interstitial if ready
            interstitialManager?.show(from: rootViewController)
        }
    }
    
    // Pragma - SASInterstitialManagerDelegate implementation
    func interstitialManager(_ manager: SASInterstitialManager, didLoad ad: SASAd) {
        delegate?.customEventInterstitialDidReceiveAd(self)
    }
    
    func interstitialManager(_ manager: SASInterstitialManager, didFailToLoadWithError error: Error) {
        delegate?.customEventInterstitial(self, didFailAd: error)
    }
    
    func interstitialManager(_ manager: SASInterstitialManager, didAppearFrom viewController: UIViewController) {
        delegate?.customEventInterstitialWillPresent(self)
    }
    
    func interstitialManager(_ manager: SASInterstitialManager, didDisappearFrom viewController: UIViewController) {
        delegate?.customEventInterstitialDidDismiss(self)
    }
    
    func interstitialManager(_ manager: SASInterstitialManager, didClickWith URL: URL) {
        delegate?.customEventInterstitialWasClicked(self)
    }
}
